src/pudl/harvest/harvest.py

```diff
@@ -0,0 +1,1054 @@
+
+        Skip harvesting by turning off the behavior in the resource metadata.
+
+        >>> resources[0]['harvest'] = False
+        >>> build = builder.build(resources)
+        >>> build['A']['x']
+        id
+        1    1
+        1    1
+        2    2
+        2    2
+        Name: x, dtype: Int64
+    """
+
+    def __init__(
+        self, dfs: Union[Dict[Any, pd.DataFrame], Iterable[pd.DataFrame]]
+    ) -> None:
+        """
+        Initialize a new ResourceBuilder.
+
+        Args:
+            dfs: Either a dictionary or iterable of dataframes.
+                If an iterable, the integer index is used as dictionary keys.
+                Dataframes may not contain duplicate column basenames
+                (e.g. 'report_year', 'report_month').
+
+        Raises:
+            ValueError: Dataframe has duplicate column basenames.
+        """
+        if not isinstance(dfs, dict):
+            # Use index as integer keys
+            dfs = {i: df for i, df in enumerate(dfs)}
+        self.dfs = dfs
+        for key in self.dfs:
+            basenames = [split_period(col)[0] for col in self.dfs[key].columns]
+            if len(set(basenames)) != len(basenames):
+                raise ValueError(f"Dataframe {key} has duplicate column basenames")
+        self.errors = None
+
+    @property
+    def columns(self) -> Set[str]:
+        """Column names of all tables."""
+        return set([name for df in self.dfs.values() for name in df.columns])
+
+    def prune(self, resources: Iterable[dict]) -> List[dict]:
+        """
+        Return only the buildable resources.
+
+        Standard resources (`harvest: False`, default) are pruned if there is no input
+        dataframe with the same name.
+        Harvested resources (`harvest: True`) are pruned if none of the primary-key
+        fields, or data fields (if any), match columns in the input dataframes. Periodic
+        key fields (e.g. 'report_month') are matched to any column of the same name with
+        an equal or smaller period (e.g. 'report_day').
+
+        Args:
+            resources: Resources with required attributes
+                `name`, `schema`.`fields`, and `schema`.`primaryKey`.
+
+        Returns:
+            The subset of `resources` which can theoretically be built.
+        """
+        pruned = []
+        for resource in resources:
+            if resource.get("harvest"):
+                columns = set([field["name"] for field in resource["schema"]["fields"]])
+                key_columns = set(resource["schema"]["primaryKey"])
+                data_columns = columns - key_columns
+                # Add alternative keys with smaller periods
+                key_columns = set(add_smaller_periods(key_columns))
+                # Keep if 1+ key columns in input, and 1+ data columns (if any) in input
+                if (self.columns & key_columns) and (
+                    not data_columns or (self.columns & data_columns)
+                ):
```

Comment:
Then you should review my harvest process description (quoted below) and tell me which logic should be changed:

> 1. Prune output tables:
> - with all key fields missing from the input tables.
> - with all (if any) data (non-key) fields missing from the input tables. For example, scenario `[a, b] -> [a*]` is unchanged, but `[a, b] -> [a*, c]` is pruned because the non-key field `c` is missing from the input table.

> 2. Test that output tables can be built:
> - All key fields appear together in one or more input tables.
> - Each data field (if any) appears in at least one of those input tables. For example, `[a] + [b] -> [a*, b]` is invalid because `b` is not present in an input table where the key fields `[a]` are also present.

My thinking was that if *all* data fields are missing from the input, then that output is not meant to be supported by those inputs and should be skipped. Alternatively, an error could be raised, which is the behavior when *some* data fields are missing from the input. But building a table with missing fields, as you are suggesting, is not how the current system is designed.

Reply:
> with all (if any) data (non-key) fields missing from the input tables. For example, scenario [a, b] -> [a*] is unchanged, but [a, b] -> [a*, c] is pruned because the non-key field c is missing from the input table.

I think this one needs to go. We need all instances of `a` no matter if it shows up with data columns or not. All other non-harvested tables will use the harvested tables for FK's. So we want `[a, b] -> [a*, c]` even if c will end up being all nulls.

> Each data field (if any) appears in at least one of those input tables. For example, [a] + [b] -> [a*, b] is invalid because b is not present in an input table where the key fields [a] are also present.

Right now, we do not dynamically generate the columns in the end-state tables. We grab the metadata which says "this table have 12 columns with 2 PK" and  use that directly. I think for consistency sake that should continue (although I haven't thought about it real hard). So given that any instance of `a` needs to end up in the final harvested table, this is still an outcome we want `[a] + [b] -> [a*, b]`, although again that `b` column should end up being nulls because there is no `b` that is directly associated with `a`.

Reply:
@cmgosnell Alright, so the only situation in which an output table is not built is when its key fields do not appear together in any of the input tables. Should that output table be skipped silently, raise an error, or return a 0-row DataFrame with all of the expected columns?

Since this means that outputs can contain columns not present in the input dataframes, the desired type of each field becomes required metadata for the harvesting process. Otherwise, there is no way to know what data type that all-null column should be.

Reply:
Yes! to your first comment.

On the first question, I think it would be best to silently skip an output table that can't.

On the latter, I think we'll want to require all columns to be dtyped in the metadata anyway so this shouldn't be a problem.